module-system/sov-state/src/codec.rs

```diff
@@ -0,0 +1,200 @@
+    /// supposed to be serializable.
+    fn encode_value(&self, value: &V) -> Vec<u8>;
+
+    /// Tries to deserialize a value from a bytes slice, and returns a
+    /// [`Result`] with either the deserialized value or an error.
+    fn try_decode_value(&self, bytes: &[u8]) -> Result<V, Self::ValueError>;
+
+    /// Deserializes a value from a bytes slice.
+    ///
+    /// # Panics
+    /// Panics if the call to [`StateValueCodec::try_decode_value`] fails. Use
+    /// [`StateValueCodec::try_decode_value`] if you need to gracefully handle
+    /// errors.
+    fn decode_value(&self, bytes: &[u8]) -> V {
+        self.try_decode_value(bytes)
+            .map_err(|err| {
+                format!(
+                    "Failed to decode value 0x{}, error: {:?}",
+                    hex::encode(bytes),
+                    err
+                )
+            })
+            .unwrap()
+    }
+}
+
+/// A market trait for types that implement both [`StateKeyCodec`] and
+/// [`StateValueCodec`].
+///
+/// This trait is blanket-implemented for all types that implement both of the
+/// above.
+pub trait StateCodec<K, V>: StateKeyCodec<K> + StateValueCodec<V> {}
+
+impl<K, V, C> StateCodec<K, V> for C where C: StateKeyCodec<K> + StateValueCodec<V> {}
+
+/// A [`StateCodec`] that uses [`borsh`] for all keys and values.
+#[derive(Debug, Default, PartialEq, Eq, Clone, borsh::BorshDeserialize, borsh::BorshSerialize)]
+pub struct BorshCodec;
+
+impl<K> StateKeyEncode<K> for BorshCodec
+where
+    K: ?Sized + borsh::BorshSerialize,
+{
+    fn encode_key(&self, key: &K) -> Vec<u8> {
+        key.try_to_vec().expect("Failed to serialize key")
+    }
+}
+
+impl<K> StateKeyCodec<K> for BorshCodec
+where
+    K: borsh::BorshSerialize + borsh::BorshDeserialize,
+{
+    type KeyError = std::io::Error;
+
+    fn try_decode_key(&self, bytes: &[u8]) -> Result<K, Self::KeyError> {
+        K::try_from_slice(bytes)
+    }
+}
+
+impl<V> StateValueCodec<V> for BorshCodec
+where
+    V: borsh::BorshSerialize + borsh::BorshDeserialize,
+{
+    type ValueError = std::io::Error;
+
+    fn encode_value(&self, value: &V) -> Vec<u8> {
+        value.try_to_vec().expect("Failed to serialize value")
+    }
+
+    fn try_decode_value(&self, bytes: &[u8]) -> Result<V, Self::ValueError> {
+        V::try_from_slice(bytes)
+    }
+}
+
+impl<K> StateKeyEncodePreservingBorrow<K, K> for BorshCodec where K: borsh::BorshSerialize {}
+impl<K> StateKeyEncodePreservingBorrow<Arc<K>, K> for BorshCodec where K: borsh::BorshSerialize {}
+impl<K> StateKeyEncodePreservingBorrow<Box<K>, K> for BorshCodec where K: borsh::BorshSerialize {}
+impl<K> StateKeyEncodePreservingBorrow<Rc<K>, K> for BorshCodec where K: borsh::BorshSerialize {}
+// Note: we're specifically *not* implementing it for `<[u8; _], [u8]>` because
+// arrays and slices have different representations in Borsh.
+impl<K> StateKeyEncodePreservingBorrow<Vec<K>, [K]> for BorshCodec where K: borsh::BorshSerialize {}
```

Comment:
Can this panic after upgrade to new borsh: https://github.com/near/borsh-rs/blob/master/borsh/src/ser/mod.rs#L275 ?